rclcpp/src/rclcpp/logging_mutex.hpp

```diff
@@ -0,0 +1,39 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__LOGGING_MUTEX_HPP_
+#define RCLCPP__LOGGING_MUTEX_HPP_
+
+#include <memory>
+#include <mutex>
+
+#include "rclcpp/visibility_control.hpp"
+
+/// Global logging mutex
+/**
+ * This mutex is locked in the following situations:
+ *   - In initialization/destruction of contexts.
+ *   - In initialization/destruction of nodes.
+ *   - In the rcl logging output handler installed by rclcpp,
+ *     i.e.: in all calls to the logger macros, including RCUTILS_* ones.
+ */
+// Implementation detail:
+// A shared pointer to the mutex is used, so that objects that need to use
+// it at destruction time can hold it alive.
+// In that way, a destruction ordering problem between static objects is avoided.
```

Comment:
That's a nice thought, but actually it will not avoid it, but instead will throw. For example, if the static in this function is destroyed and then you attempt to shutdown logging, an exception is thrown instead.

Reply:
Instead we may need to have logic that allows shutdown to happen without a mutex lock in the event that this function returns nullptr? And also causes log functions to fail without trying to log if this returns nullptr? That should allow this to be destroyed first, and then at least let logging shutdown gracefully-ish.

Reply:
> That's a nice thought, but actually it will not avoid it, but instead will throw. For example, if the static in this function is destroyed and then you attempt to shutdown logging, an exception is thrown instead.

None of the destructors (NodeBase/Context) is calling `get_global_logging_mutex`. This function is only being called at construction time, ensuring that at that point the static shared pointer will be alive.

In all places where the mutex is needed at destruction time, a copy of the shared pointer that was made a construction time is being used. That copy of the shared pointer is hold by the object that needs to lock the mutex in its destructor, ensuring it will still be alive (if the shared pointer with static lifetime in `get_global_logging_mutex` was already destructed, that won't be a problem as another owner is keeping the mutex alive).

I cannot see the problem you're pointing out. If you think this is a problem, please be more specific in how the use-after-destruction can happen.

Reply:
Sorry, I got bamboozled by misreading the github view, haha:

![image](https://user-images.githubusercontent.com/100427/82700475-1e0a4780-9c23-11ea-9e8e-acd23394d27b.png)

I thought it was being used in the context destructor.

You're right, I think this ought to work then.